protocols/kad/src/async_behaviour.rs

```diff
@@ -0,0 +1,346 @@
+use std::{collections::HashMap, task::Poll};
+
+use futures::channel::mpsc;
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour, THandler, THandlerInEvent,
+    THandlerOutEvent, ToSwarm,
+};
+
+use crate::{
+    kbucket, store, AddProviderResult, Behaviour, BootstrapResult, Event, GetClosestPeersResult,
+    GetProvidersResult, GetRecordResult, NoKnownPeers, PutRecordResult, QueryId, QueryResult,
+    QueryStats, Quorum, Record, RecordKey,
+};
+
+/// The results of Kademlia queries (strongly typed) and only
+/// those initiated by the user.
+pub struct AsyncQueryResult<T> {
+    pub id: QueryId,
+    pub result: T,
+    pub stats: QueryStats,
+}
+impl<T> AsyncQueryResult<T> {
+    fn map<Out>(self, f: impl Fn(T) -> Out) -> AsyncQueryResult<Out> {
+        AsyncQueryResult {
+            id: self.id,
+            stats: self.stats,
+            result: f(self.result),
+        }
+    }
+}
+
+type UnboundedQueryResultSender<T> = mpsc::UnboundedSender<AsyncQueryResult<T>>;
```

Comment:
Yeah you're right. I'm going to introduce a wrapper `AsyncQueryResultStream` around `UnboundedQueryResultReceiver` which implement `futures::Stream`. It will still depends on the definition of `futures::Stream` but I don't  think it is possible to remove it and since it is done in `protocols/stream/src/control.rs` for `IncomingStreams` I think it is ok. 

https://github.com/libp2p/rust-libp2p/blob/9bb2a871939ded657d380cc2b7beb519840f3e8d/protocols/stream/src/control.rs#L127-L133

Is it ok ?

Reply:
I'm closing this since I think only depending on the definition of `futures::Stream` is ok. If you don't agree, feel free to reopen it.